### PR TITLE
mon: do not panic at seeing requests without session

### PR DESCRIPTION
**Pull request by tchaikov**
http://tracker.ceph.com/issues/15113

### CODE REVIEW

**Comment by tchaikov** [comment]
@jecluis mind review it?

File: src/mon/Monitor.cc [hunk 1]
```diff
@@ -3329,40 +3329,63 @@ void Monitor::handle_route(MonOpRequestRef op)
   }
 }
 
+namespace {
+  struct ReentrantLock {
+    const bool initially_locked;
+    Mutex &mutex;
+    ReentrantLock(Mutex& m)
+      : initially_locked(m.is_locked_by_me()),
```
Comment (liewegas) [inline src/mon/Monitor.cc]: I don't think we should ever use is_locked* as a conditional--only for debug assertions.  Notably, this is not available from std::mutex.

Comment (liewegas) [inline src/mon/Monitor.cc]: Better to make the caller(s) who don't already hold lock take it before calling resend_routed_requests?

Comment (tchaikov) [inline src/mon/Monitor.cc]: so we have following call stack:

Monitor::init() => with lock: bootstrap() => win_standalone_election() => win_election() => Monitor::finish_election() => Monitor::resend_routed_requests()

Elector::expire() => Elector::victory() => **with lock:** Monitor::win_election() => Monitor::finish_election() => Monitor::resend_routed_requests()

Elector::expire() => **with lock:** bootstrap() => win_standalone_election() => win_election() => Monitor::finish_election() => Monitor::resend_routed_requests()

the "with lock:" in bold font are added in this PR.

**Comment by tchaikov** [comment]
changelog

- rework to avoid using "is_locked_by_me"
- rebase against master.

not tested, so "[DNM]".

**Comment by tchaikov** [comment]
changelog

- remove the commit to hold the mon lock when resend_routed_requests(): the safe timer does hold the lock for us, so it's not necessary and actually wrong!
- rebased against master.

**Comment by liewegas** [comment]
passed rados suite

**Comment by tchaikov** [comment]
@jecluis ping?

**Comment by jecluis** [comment]
LGTM

**Pull request merged**